### ISSUE TITLE
New Step 'Template'

### ISSUE THREAD

**Issue by gaughan** [syndesisio/syndesis#3047]
## Feature Owner: Paul

As a citizen integrator, I want to be able to add a Template in a step connector.
Each Template step must have a template added. FreeMarker and Velocity and Mustache should be supported.
In the step configuration, I want to be able to add via file upload or URL. I can choose a type (FreeMarker) or preferably this is automatically set by reading the file. 
Once a file loads display a text editor, with validate and inline context editor?
Once a template step is complete, a warning is displayed (shape is present) and data mapping is now required.
Example template for illustration:
Example.tmpl
```freemarker
Hello $name
Thanks for buying $product
Cheers
$company
```

In the above example, if you return to change the product name to product2 does it break the mapper? Can we catch this? (As per the call, yes this should be ok)

Nice to have:
Ability to edit the template later

DoD
- Mustache is the only supported format for 7.2
- Simple text editor is available.
- No inline validation is required for 7.2
- Namespace support will be provided and the chosen namespace format($header.product ?) should work with FreeMarker and Velocity and Mustache

**Comment by amysueg** [comment]
Twitter (start; typed example)
-map-
Template step (specifies which parameters)
Gmail (finish; since unstructured data type, accepts anything, no mapping step required)

If the finish is "Salesforce", as a typed connector it will in most cases require a mapping step from the Template step.

**Comment by TovaCohen** [comment]
Thanks for the notes, Amy! These template things are new to me. So I looked them up. Here's a short description of the three template implementations that should initially be supported. I'm adding this here in case anyone else might find this useful. 

- Apache **FreeMarker**™ is a template engine: a Java library to generate text output (HTML web pages, e-mails, configuration files, source code, etc.) based on templates and changing data. Templates are written in the FreeMarker Template Language (FTL), which is a simple, specialized language (not a full-blown programming language like PHP). 

- Apache **Velocity** is a Java-based template engine. It permits anyone to use a simple yet powerful template language to reference objects defined in Java code.

- **Mustache** is a logic-less template syntax. It can be used for HTML, config files, source code - anything. It works by expanding tags in a template using values provided in a hash or object. Some tags are replaced with a value, some nothing, and others a series of values. 

A "Template" step is the kind of step that operates on data between connections.

**Comment by phantomjinx** [comment]
@amysueg @TovaCohen 

> Twitter (start; typed example)
> -map-
> Template step (specifies which parameters)
> Gmail (finish; since unstructured data type, accepts anything, no mapping step required)

Can you confirm that the expected output of the Template step is a variable/parameter such as $text or $body? The GMail connector in turn takes a set of parameters, ie. emailTo, emailSubject, emailText. Would it then be the case that emailText would be assigned $text/$body?

Given you suggested that the Gmail connector is unstructured, the alternative would be that the Gmail connector no longer needs to display any of these parameters in the UI since all of that should have been added in the Step Template and therefore emailTo, emailSubject etc.. are redundant?

Thanks.

**Comment by gaughan** [comment]
@phantomjinx 
Template step is generic (not coupled to gmail)
No changes to gmail connector are expected.
Namespace creation if required should be applicable across multiple template languages.

**Comment by phantomjinx** [comment]
@gaughan
Thanks for that. So just trying to get a handle on what the output coming out of the template step is likely to be and how the next step/connection in the workflow of the integration is going to exploit it.

**Comment by gaughan** [comment]
Also nice to have the ability to edit the template later.
